src/datatypes/src/vectors/decimal.rs

```diff
@@ -0,0 +1,465 @@
+                Value::Decimal128(Decimal128::new_unchecked(value, *precision, *scale))
+            }
+            _ => Value::Null,
+        }
+    }
+
+    fn get_ref(&self, index: usize) -> ValueRef {
+        if !self.array.is_valid(index) {
+            return ValueRef::Null;
+        }
+
+        match self.array.data_type() {
+            ArrowDataType::Decimal128(precision, scale) => {
+                // Safety: The index have been checked by `is_valid()`.
+                let value = unsafe { self.array.value_unchecked(index) };
+                ValueRef::Decimal128(Decimal128::new_unchecked(value, *precision, *scale))
+            }
+            _ => ValueRef::Null,
+        }
+    }
+}
+
+impl From<Decimal128Array> for Decimal128Vector {
+    fn from(array: Decimal128Array) -> Self {
+        Self { array }
+    }
+}
+
+impl From<Vec<Option<i128>>> for Decimal128Vector {
+    fn from(vec: Vec<Option<i128>>) -> Self {
+        let array = Decimal128Array::from_iter(vec);
+        Self { array }
+    }
+}
+
+impl Serializable for Decimal128Vector {
+    fn serialize_to_json(&self) -> Result<Vec<serde_json::Value>> {
+        self.iter_data()
+            .map(|v| match v {
+                None => Ok(serde_json::Value::Null), // if decimal vector not present, map to NULL
+                Some(vec) => serde_json::to_value(vec),
+            })
+            .collect::<serde_json::Result<_>>()
+            .context(error::SerializeSnafu)
+    }
+}
+
+pub struct Decimal128Iter<'a> {
+    data_type: &'a ArrowDataType,
+    iter: ArrayIter<&'a Decimal128Array>,
+}
+
+impl<'a> Iterator for Decimal128Iter<'a> {
+    type Item = Option<Decimal128>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        Some(self.iter.next().and_then(|v| {
+            v.and_then(|v| match self.data_type {
+                ArrowDataType::Decimal128(precision, scale) => {
+                    Some(Decimal128::new_unchecked(v, *precision, *scale))
+                }
+                _ => None,
+            })
+        }))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+impl ScalarVector for Decimal128Vector {
+    type OwnedItem = Decimal128;
+
+    type RefItem<'a> = Decimal128;
+
+    type Iter<'a> = Decimal128Iter<'a>;
+
+    type Builder = Decimal128VectorBuilder;
+
+    fn get_data(&self, idx: usize) -> Option<Self::RefItem<'_>> {
+        if !self.array.is_valid(idx) {
+            return None;
+        }
+
+        match self.array.data_type() {
+            ArrowDataType::Decimal128(precision, scale) => {
+                // Safety: The index have been checked by `is_valid()`.
+                let value = unsafe { self.array.value_unchecked(idx) };
+                // Safety: The precision and scale have been checked by ArrowDataType.
+                Some(Decimal128::new_unchecked(value, *precision, *scale))
+            }
+            _ => None,
+        }
```

Comment:
Can we provide a helper function to get an `Option<Decimal128>` from the array? I see similar logic repeated many times. In `get`/`get_ref` we can use into to cast it to `Value`/`ValueRef`